src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj

```diff
@@ -299,9 +300,10 @@
                      (:id instance)
                      max-requests)
           (flush-instance-fn pool instance))
-        (.putFirst pool instance)))
-    ;; if we get here, we got a Retry, so we just put it back into the pool.
-    (.putFirst (:pool instance) instance)))
+        (.returnItem pool instance)))
+    ;; if we get here, it was from a borrow and we got a Retry, so we just
+    ;; return it to the pool.
+    (.returnItem (:pool instance) instance)))
```

Comment:
I spent a while trying to puzzle over this, and I think (but am not convinced) that this needs to be a `returnItem`. I changed the `putFirst` in `flush-pool!` to `insertPill` (https://github.com/puppetlabs/puppet-server/pull/698/files#diff-af969cd26304257b62d766cb4c5325faR118), because there there isn't anything holding a read lock; we're just adding the `RetryPoisonPill` to the pool for the first time.

Here, on the other hand, it's possible that a borrow could have gotten the retry pill (https://github.com/puppetlabs/puppet-server/blob/master/src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj#L106). In that case, since it did a borrow it holds a read lock, which we need to unlock, which is why `returnItem`.